Adding Synchronization between PM UI and NuGet project Cache

mishra14 opened the pull request:
Fixes: https://github.com/NuGet/Home/issues/4035 for vs 2017 and other dupe bugs.

Currently, the Package Manager UI has no synchronization with project cache. Main problem here comes on .NET Core projects which invoke restore asynchronously and thus the UI can be left out of date.

As a solution I have added 2 step event synchronization. The `ProjectSystemCache.cs` class fires a `CacheUpdate` Event when cache is updated i.e. `ProjectRestoreInfoAdded`, `ProjectAdded`, `ProjectRemoved `and `Clear`. The Event is consumed by `VSSolutionManager `which then asynchronously fires another event i.e. `AfterNuGetCacheUpdated`. This event is then consumed by `PackageManagerControl.xaml.cs` to refresh the UI.

//cc: @alpaix @jainaashish @emgarten @rrelyea @rohit21agrawal @nkolev92 @zhili1208

Commit:
Adding events

Commit:
fixing build error

Commit:
Fixing bug on cache check

Commit:
removing UI event on close

Commit:
Adding tests

Commit:
Fixed a syntax error

emgarten commented on src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs:
```diff
@@ -25,6 +29,20 @@ internal class ProjectSystemCache : IProjectSystemCache
         private readonly Dictionary<string, HashSet<ProjectNames>> _shortNameCache = new Dictionary<string, HashSet<ProjectNames>>(StringComparer.OrdinalIgnoreCase);
 
+        /// <summary>
+        /// Returns the current value of _isCacheDirty.
```
This could be more helpful :smile:

mishra14 replied:
Hence - https://github.com/NuGet/NuGet.Client/pull/1152/files/d1b82d2b3e99089345f53035e956d7d256223665#diff-c38fac88f9b4a4ed4bfb16060e700ed7R22

:p

emgarten replied:
put that comment on the property maybe. /// comments go into the intellisense and are much more visible

alpaix commented on src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -1004,7 +1009,32 @@ internal static IEnumerable<EnvDTE.Project> GetDependentEnvDTEProjects(IDictiona
-#region IVsSelectionEvents
+        private void NuGetCacheUpdate_After(object sender, EventArgs e)
+        {
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(() => FireNuGetCahceUpdatedEventAsync(e));
```
Please add a comment explaining why you're doing this.

emgarten replied:
Typo: Cahce

emgarten commented on src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -1004,7 +1009,32 @@ internal static IEnumerable<EnvDTE.Project> GetDependentEnvDTEProjects(IDictiona
+                // Await a delay of 100 mSec to batch multiple cache updated events
+                await Task.Delay(100);
+                // Check if the cache is still dirty
+                if (_projectSystemCache.TestResetDirtyFlag())
```
This check seems like it would be very cheap, should this also be done before starting a new thread?

mishra14 replied:
discussed offline. This prevents multiple invocation of UI  refresh.

emgarten commented on src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -1004,7 +1009,32 @@ internal static IEnumerable<EnvDTE.Project> GetDependentEnvDTEProjects(IDictiona
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex.ToString());
```
log the exception at the debug level or to the activity log. Displaying a stack trace to the user and not actually stopping the action seems incorrect.

alpaix replied:
`_logger` here DOES actually write to ActivityLog

alpaix commented on src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs:
```diff
@@ -130,6 +148,8 @@ public void RemoveProject(string name)
                 _readerWriterLock.ExitUpgradeableReadLock();
             }
+
+            FireCacheUpdatedEvent();
```
We should probably fire the event only in `AddProjectRestoreInfo` which is called via the nomination. This'd be just enough to resolve the bug. Not sure what side effects will be with so many events raised.

alpaix commented on src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectSystemCache.cs:
```diff
@@ -459,5 +486,37 @@ private class CacheEntry
+            // We should fire the event only if the cache was clean before.
+            // If the cache was dirty already then the VSSolutionsManager is yet to consume the event
+            // and will consume current changes as well.
+            if(CacheUpdated!= null && TestSetDirtyFlag())
```
nit: format documnet

emgarten replied:
nit: formatting

emgarten commented on src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/IProjectSystemCache.cs:
```diff
@@ -107,5 +114,11 @@ public interface IProjectSystemCache
+        /// Reset the dirty flag to 0 (is Not Dirty) if the flag was already set ( like TSL!!).
+        /// </summary>
+        /// <returns><code>true</code> if the cache was dirty before and <code>false</code> otherwise</returns>
+        bool TestResetDirtyFlag();
```
Not sure I get the TSL reference here. The name seems confusing to me. GetAndResetDirtyFlag?

mishra14 replied:
Sorry, the comment was not updated in the interface.

alpaix commented on src/NuGet.Core/NuGet.PackageManagement/IDE/ISolutionManager.cs:
```diff
@@ -18,6 +18,8 @@ public interface ISolutionManager
 
         event EventHandler SolutionClosed;
 
+        event EventHandler AfterNuGetCacheUpdated;
```
move it to `IVsSolutionManager` to avoid touching core interface

mishra14 replied:
Discussed offline. Moving it to IVsSolutionmanager would require chnages to NuGetUIContext and that would require changes to anyone that consumes it.

emgarten reviewed (changes requested):
comments

alpaix commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/ProjectSystemCacheTests.cs:
```diff
@@ -193,5 +194,266 @@ public void AddProject_AfterAddProjectRestoreInfo_UpdatesCacheEntry()
         }
+
+        [Fact]
+        public void UpdatingCacheTriggersEvent_NoEventHandler()
```
First portion of a test name usually is a name of the method under test (Act).

Commit:
Addressing PR feedback

mishra14 commented:
🔔

jainaashish commented on src/NuGet.Clients/PackageManagement.VisualStudio/IDE/IVsSolutionManager.cs:
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
```
why this added?

jainaashish commented on src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -1004,7 +1009,40 @@ internal static IEnumerable<EnvDTE.Project> GetDependentEnvDTEProjects(IDictiona
+        {
+            // The AfterNuGetCacheUpdated event is raised on a separate Task to prevent blocking of the caller.
+            // E.g. - If Restore updates the cache entries on CPS nomination, then restore should not be blocked till UI is restored.
+            NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(() => FireNuGetCacheUpdatedEventAsync(e));
```
why dont you use `Task.Run` instead of `JTF.RunAsync`? any specific reason for that?

emgarten replied:
I looked at this with @mishra14 earlier, it can go on to the UI thread deeper into the call, but it is in a different JTF then. Task.Run could be used here I think, but it is unclear if it would make a difference. @jainaashish do you know if there is a perf difference?

mishra14 replied:
@jainaashish : as @emgarten  mentioned, what would be the benefit of Task.Run over JTF?

jainaashish replied:
I don't think there would be any significant perf difference between these two. But I would recommend to use Task.Run over JTF any day. Unless it's required to use JTF, don't use it and in your case Task,Run will work fine unless I'm missing something

alpaix replied:
Using JTF helps avoiding deadlocks in VS. As a rule we should NOT create tasks with `Task.Run` while in VS level code.

jainaashish replied:
And if not used properly then JTF is also the root for any deadlocks in VS, what do you see wrong with Task.Run as in this case it's like fire and forget about this task so don't you think it will be a better approach?

jainaashish replied:
discussed offline, it's good to use JTF here

jainaashish commented on src/NuGet.Clients/PackageManagement.VisualStudio/IDE/VSSolutionManager.cs:
```diff
@@ -1004,7 +1009,40 @@ internal static IEnumerable<EnvDTE.Project> GetDependentEnvDTEProjects(IDictiona
+            try
+            {
+                // Await a delay of 100 mSec to batch multiple cache updated events
+                await Task.Delay(100);
```
what is the basis of this 100ms?

emgarten replied:
💯 ✨

mishra14 replied:
This ensures the minimum duration between 2 consecutive UI refresh, caused by cache update, to be 100 mSec. No "scientific" reason if that's what you are asking.

jainaashish reviewed (changes requested):
Few comments to address. Also this issue reproduces when we switch between tabs (browse, install, update) will this PR also address that?

And I would say to add tests with TestSolutionManager to atleast receive events from ProjectSystemCache

emgarten reviewed (approved):
Discussed most of these offline, the change looks good.

rohit21agrawal commented on src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/IProjectSystemCache.cs:
```diff
@@ -107,5 +121,12 @@ public interface IProjectSystemCache
+        /// This is public so that external callers can inform the cache that they have consumed the updated cache event.
+        /// </summary>
+        /// <returns><code>true</code> if the cache was dirty before and <code>false</code> otherwise</returns>
+        bool TestResetDirtyFlag();
```
why TestResetDirtyFlag instead of ResetDirtyFlag?

mishra14 replied:
Because the method resets the DirtyFlag only if the flag was set before hence Test Prefix. it behaves like a Test-Set Lock/TSL.

mishra14 commented:
@jainaashish : This will address the issue with all the UI tabs. 
I will look into adding test cases with TestSolutionManager.

rohit21agrawal commented on src/NuGet.Clients/PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -146,6 +146,7 @@ private PackageSourceMoniker SelectedSource
             solutionManager.NuGetProjectUpdated += SolutionManager_ProjectsUpdated;
             solutionManager.NuGetProjectRenamed += SolutionManager_ProjectRenamed;
             solutionManager.ActionsExecuted += SolutionManager_ActionsExecuted;
+            solutionManager.AfterNuGetCacheUpdated += SolutionManager_CacheUpdated;
```
I think instead of creating a new SolutionManager_CacheUpdated function, SolutionManager_ActionsExecuted function can be reused.

Commit:
Adding code to pass project name as an event arg

Commit:
Adding E2E test to check VSSolutionManager event raising + clean up

mishra14 commented:
🔔 @rohit21agrawal @jainaashish @alpaix @emgarten : Changes - 

1. Updated the EventArgs to NUGetEventArgs to allow passing project.FullName. This limits the UI refresh to cases where the UI is opened for the solution or the project for which the cache was updated.
2. Added an E2E test to assert that the solution manager fires an event in case of package install.

rohit21agrawal commented on src/NuGet.Core/NuGet.PackageManagement/Utility/EventArgs.cs:
```diff
@@ -0,0 +1,33 @@
+    /// Generic EventArg class for NuGet.
+    /// </summary>
+    /// <typeparam name="T">Class type of the argument Arg.</typeparam>
+    public sealed class NuGetEventArgs<T> : EventArgs where T : class
```
rename file to NuGetEventArgs to be consistent.

emgarten reviewed (approved):
Fix filename then :shipit:

alpaix commented on src/NuGet.Clients/PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -221,6 +222,33 @@ private void SolutionManager_ActionsExecuted(object sender, ActionsExecutedEvent
+            else
+            {
+                // This is a project package manager, so there is one and only one project.
+                var project = Model.Context.Projects.First();
```
use `.Single()` then

mishra14 replied:
discussed offline. Keeping this since it is the same as the code for `SolutionManager_ActionsExecuted`.

alpaix reviewed (approved):
Looks :sparkles: to me!

rohit21agrawal commented:
:shipit:
